File: myeventbus/src/main/java/com/example/chimu/myeventbus/MainActivity.java
package com.example.chimu.myeventbus;

import android.content.Intent;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

public class MainActivity extends AppCompatActivity {
    private Button button2;
    private String Msg;
    private TextView textView;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        button2 = (Button) findViewById(R.id.button2);

        textView = (TextView) findViewById(R.id.msg);
        button2.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(MainActivity.this,OtherActivity.class);
                startActivity(intent);
            }
        });

        //订阅
        EventBus.getDefault().register(this);


    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        //取消订阅
        EventBus.getDefault().unregister(this);
    }


/**
 * ThreadMode总共四个：
 NAIN UI主线程
 BACKGROUND 后台线程
 POSTING 和发布者处在同一个线程
 ASYNC 异步线程
 *
 * */
    //发布事件的处理
    @Subscribe(threadMode = ThreadMode.MAIN,priority = 100)   //UI线程执行，优先级100，优先级和广播类似
    public void  onMyEvent(String  msg){

        textView.setText(msg);
        Log.i("msg","msg:"+msg);
        // EventBus.getDefault().cancelEventDelivery(myEvent) ;//优先级高的订阅者可以终止事件往下传递
    }


}

File: kotlin/src/main/java/com/example/chimu/kotlin/MainActivity.kt
package com.example.chimu.kotlin

import android.content.Intent
import android.support.v7.app.AppCompatActivity
import android.os.Bundle

class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

//        println(getName(2))
//        println(sum(3, 5))
//        println("isString" + isString(1));
//        println(isNull());

        //函数式编程FP
//        val numbers = listOf(1,2,3)
//        print(numbers.filter(::isOdd))//有问题

        //lambda 表达式
        println(fsum0(1, 1))
        println(fsum0.invoke(1,1))

        println(fsum1(1, 1).invoke())
        println(fsum2(1, 1))
        println(1.fsum3(1))
    }
    private fun getName(id:Int): String{
        return "迟暮"
    }
    fun sum(a: Int, b: Int): Int { // kotlin中的返回值类型必须明确指定
        println("sum of $a and $b is ${a + b}")
        return a + b
    }
    fun isString(a:Any):Boolean{
        var isS  :Boolean
        if(a is String) isS = true else isS = false
        return isS
    }

    fun isNull(){
        val s: String? = null
        println(s?.length)
    }
    fun isOdd(x:Int) = x % 2 != 0

    //    Lambda 表达式
    val fsum0 = { x: Int, y: Int ->  x + y  }
    val fsum1 = { x: Int, y: Int -> { x + y } }
    val fsum2 = fun(x: Int, y: Int): Int = x + y
    val fsum3 = fun Int.(other: Int): Int = this + other
}
